File: Tests/CeyradTests/SourceSelectorTests.swift
import XCTest

@testable import Ceyrad

final class SourceSelectorTests: XCTestCase {
    private func source(
        running: Bool = true, state: PlayerState = .stopped,
        hasTrack: Bool = true
    ) -> SourceState {
        var s = SourceState()
        s.running = running
        s.playerState = state
        s.track = hasTrack ? TrackInfo(name: "T", artist: "A", album: "L") : nil
        return s
    }

    private func select(_ am: SourceState, current: MusicSourceID? = nil) -> MusicSourceID? {
        SourceSelector.selectActiveSource(appleMusic: am, current: current)
    }

    func testNothingRunningYieldsNil() {
        XCTAssertNil(select(source(running: false, hasTrack: false)))
    }

    func testPlayingIsACandidate() {
        XCTAssertEqual(select(source(state: .playing)), .appleMusic)
    }

    func testPausedIsStillACandidate() {
        // 一時停止も「表示すべき候補」ではある（消すかどうかはActivityBuilder側の判断）
        XCTAssertEqual(select(source(state: .paused)), .appleMusic)
    }

    func testStoppedSourceIsNotACandidate() {
        XCTAssertNil(select(source(state: .stopped, hasTrack: false)))
    }

    func testRunningWithoutTrackIsNotACandidate() {
        // 稼働していても曲情報がなければ表示しない（初期状態取得前など）
        XCTAssertNil(select(source(state: .playing, hasTrack: false)))
    }

    func testNotRunningIsNotACandidate() {
        XCTAssertNil(select(source(running: false, state: .playing)))
    }
}

File: windows/tests/source_selector_tests.rs
//! Port of Tests/CeyradTests/SourceSelectorTests.swift

use ceyrad::core::models::{MusicSourceId, PlayerState, SourceState, TrackInfo};
use ceyrad::core::source_selector::select_active_source;

fn source(running: bool, state: PlayerState, has_track: bool) -> SourceState {
    SourceState {
        running,
        player_state: state,
        track: if has_track {
            Some(TrackInfo::new("T", "A", "L"))
        } else {
            None
        },
        ..SourceState::default()
    }
}

#[test]
fn nothing_running_yields_none() {
    let not_running = source(false, PlayerState::Stopped, false);
    assert_eq!(select_active_source(&not_running), None);
}

#[test]
fn a_playing_source_wins() {
    let playing = source(true, PlayerState::Playing, true);
    assert_eq!(
        select_active_source(&playing),
        Some(MusicSourceId::AppleMusic)
    );
}

#[test]
fn a_paused_source_still_counts() {
    // Keep showing the presence while merely paused, so it doesn't flip off.
    let paused = source(true, PlayerState::Paused, true);
    assert_eq!(
        select_active_source(&paused),
        Some(MusicSourceId::AppleMusic)
    );
}

#[test]
fn stopped_source_is_not_a_candidate() {
    let stopped = source(true, PlayerState::Stopped, true);
    assert_eq!(select_active_source(&stopped), None);
}

#[test]
fn running_without_track_is_not_a_candidate() {
    let no_track = source(true, PlayerState::Playing, false);
    assert_eq!(select_active_source(&no_track), None);
}
